core/state/state_object.go

```diff
@@ -589,14 +574,10 @@ func (s *stateObject) CodeSize() int {
 }
 
 func (s *stateObject) SetCode(codeHash common.Hash, code []byte) {
-	prevcode := s.Code()
-	s.db.journal.append(codeChange{
-		account:  &s.address,
-		prevhash: s.CodeHash(),
-		prevcode: prevcode,
-	})
+	s.db.journal.SetCode(s.address)
```

Comment:
This change will implicitly change the function semantics, which now always assumes the code was empty.

It's actually not true in some circumstances, e.g. StateOverride in RPC level might rewrites the existent code to another, the code will be wrong if rollback happens.

Reply:
Also, from the statedb's perspective, it's valid to change code from one value to another. The restriction(the code must be empty before setting) is at a higher level. 

I think it's better to keep the original behavior, it's more self-contained.

Reply:
![Screenshot 2024-08-23 at 09-21-12 core_state semantic journalling (part 1) by holiman · Pull Request #28880 · ethereum_go-ethereum](https://github.com/user-attachments/assets/5ab664dd-233f-42ef-a0d5-8a30fd13b582)

Yes, I wrote about this on the PR description. 

The drawback retaining the old semantics is that the old code needs to be "remembered" in the journal-entry, using a `prevcode`. In practice, `prevcode` is always empty. Except for maybe in some cornercases around testing, where the ethereum protocol rules are not used. 
 
For StateOverride, we don't actually have to journal the setting of code. We won't ever roll back that change, will we?

Reply:
> The drawback retaining the old semantics is that the old code needs to be "remembered" in the journal-entry,

In practice, the old contract code will always be zero, so it's basically "no-cost" to memorize in journal?

Reply:
> For StateOverride, we don't actually have to journal the setting of code. We won't ever roll back that change, will we?

True. But just for the example, my concern is that we might support some weird features/operations in the future and changing code might be possible. At that time, the revert logic is wrong.

Reply:
>  But just for the example, my concern is that we might support some weird features/operations in the future and changing code might be possible. At that time, the revert logic is wrong.

If that's an issue, which I think it never will be, we could hack it pretty easily. We could e.g. add a `DeleteCode` which sets code back to empty, followed by a `SetCode`. The `DeleteCode` could remember the previous code. 

And we can use the same thing if we want to journal StateOverride. 

> In practice, the old contract code will always be zero, so it's basically "no-cost" to memorize in journal?

True, the resource-overhead of holding an extra pointer is pretty small, but I think the code complexity goes down by making this change. It removes a struct-field, and makes the code IMO more un-ambiguous. I mean, the code becomes less generic, and more specific.